reactor-core/src/main/java/reactor/core/publisher/FluxRefCount.java

```diff
@@ -61,21 +57,58 @@ public int getPrefetch() {
 	@Override
 	public void subscribe(CoreSubscriber<? super T> actual) {
 		RefCountMonitor<T> state;
-		
-		for (;;) {
+		boolean connect = false;
+
+		synchronized (this) {
 			state = connection;
-			if (state == null || OperatorDisposables.isDisposed(state.disconnect)) {
-				RefCountMonitor<T> u = new RefCountMonitor<>(n, this);
-				
-				if (!CONNECTION.compareAndSet(this, state, u)) {
-					continue;
+			if (state == null) {
+				state = new RefCountMonitor<>(this);
+				connection = state;
+			}
+
+			long count = state.subscribers;
+			state.subscribers = count + 1;
+			if (!connection.connected && count + 1 == n) {
+				connect = true;
+				state.connected = true;
+			}
+		}
+
+		source.subscribe(new RefCountInner<>(actual, state));
+		if (connect) {
+			source.connect(state);
+		}
+	}
+
+	void connectionCancel(RefCountMonitor rc) {
+		synchronized (this) {
```

Comment:
we can't use synchronized if we nest user/external calls (like dispose())

Reply:
do you mean we need to exit the synchronized block before calling the user code? (this reflects the way things are done in RxJava2 btw)

Reply:
external call poses the risk that the `dispose()` method would recursively re-subscribe, need to polish that